skyportal/handlers/thumbnail.py

```diff
@@ -0,0 +1,136 @@
+            source = Source.query.get(data['source_id'])
+            try:
+                phot = source.photometry[0]
+            except IndexError:
+                return self.error('Specified source does not yet have any photometry data.')
+        else:
+            return self.error('One of either source_id or photometry_id are required.')
+        try:
+            t = create_thumbnail(data['data'], data['ttype'], source.id, phot)
+        except Exception as e:
+            return self.error(str(e))
+        DBSession().commit()
+
+        return self.success(data={"id": t.id})
+
+    @auth_or_token
+    def get(self, thumbnail_id):
+        info = {}
+        info['thumbnail'] = Thumbnail.query.get(thumbnail_id)
+
+        if info['thumbnail'] is not None:
+            return self.success(data=info)
+        else:
+            return self.error(f"Could not load thumbnail {thumbnail_id}",
+                              data={"thumbnail_id": thumbnail_id})
+
+    @permissions(['Manage sources'])
+    def put(self, thumbnail_id):
+        data = self.get_json()
+        data['id'] = thumbnail_id
+
+        schema = Thumbnail.__schema__()
+        try:
+            schema.load(data)
+        except ValidationError as e:
+            return self.error('Invalid/missing parameters: '
+                              f'{e.normalized_messages()}')
+        DBSession().commit()
+
+        return self.success()
+
+    @permissions(['Manage sources'])
+    def delete(self, thumbnail_id):
+        DBSession.query(Thumbnail).filter(Thumbnail.id == int(thumbnail_id)).delete()
+        DBSession().commit()
+
+        return self.success()
+
+
+def create_thumbnail(thumbnail_data, thumbnail_type, source_id, photometry_obj):
+    basedir = Path(os.path.dirname(__file__))/'..'/'..'
+    if os.path.abspath(basedir).endswith('skyportal/skyportal'):
+        basedir = basedir/'..'
+    file_uri = os.path.abspath(
+        basedir/f'static/thumbnails/{source_id}_{thumbnail_type}.png')
+    if not os.path.exists(os.path.dirname(file_uri)):
+        (basedir/'static/thumbnails').mkdir(parents=True)
+    file_bytes = base64.b64decode(thumbnail_data)
+    im = Image.open(io.BytesIO(file_bytes))
+    if im.format != 'PNG':
+        raise ValueError('Invalid thumbnail image type. Only PNG are supported.')
+    if not (100, 100) <= im.size <= (500, 500):
+        raise ValueError('Invalid thumbnail size. Only thumbnails '
+                        'between (100, 100) and (500, 500) allowed.')
+    try:
+        t = Thumbnail(type=thumbnail_type,
+                      photometry=photometry_obj,
+                      file_uri=file_uri,
+                      public_url=f'/static/thumbnails/{source_id}_{thumbnail_type}.png')
+        DBSession.add(t)
+        DBSession.flush()
+    except LookupError:
+        raise LookupError('Invalid thumbnail type. Please refer to '
+                          'API docs for a list of allowed types.')
+    except Exception as e:
```

Comment:
Hrm, I guess we could also just remove the "except" clause here with the same effect.